<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/botanicgrasping/workspace/botanic-grasping/install/open_manipulator_pro_description/share/open_manipulator_pro_description/urdf/open_manipulator_pro.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="open_manipulator_pro">
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/botanicgrasping/workspace/botanic-grasping/install/open_manipulator_pro_control/share/open_manipulator_pro_control/config/controllers.yaml</parameters>
    </plugin>
  </gazebo>
  <!-- <gazebo>
      <plugin name="gazebo_ros_joint_state_publisher"
          filename="libgazebo_ros_joint_state_publisher.so">
          <update_rate>100</update_rate>
          <joint_name>joint1</joint_name>
          <joint_name>joint2</joint_name>
          <joint_name>joint3</joint_name>
          <joint_name>joint4</joint_name>
          <joint_name>joint5</joint_name>
          <joint_name>joint6</joint_name>
      </plugin>
  </gazebo> -->
  <!--
  <gazebo>
      <plugin name="gazebo_ros_joint_pose_trajectory"
          filename="libgazebo_ros_joint_pose_trajectory.so">
          <update_rate>2</update_rate>
      </plugin>
  </gazebo> -->
  <gazebo reference="camera_link">
    <material>Gazebo/Red</material>
    <!-- <sensor name="camera" type="camera"> -->
    <sensor name="camera" type="depth">
      <pose> 0 0 0 0 0 0 </pose>
      <visualize>true</visualize>
      <update_rate>10</update_rate>
      <camera>
        <horizontal_fov>1.089</horizontal_fov>
        <image>
          <format>R8G8B8</format>
          <width>640</width>
          <height>480</height>
          <!-- <width>1280</width>
                  <height>960</height> -->
        </image>
        <clip>
          <near>0.01</near>
          <far>8.0</far>
        </clip>
      </camera>
      <plugin filename="libgazebo_ros_camera.so" name="camera_controller">
        <frame_name>camera_link_optical</frame_name>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
    <!-- <sensor name="camera" type="depth">
          <visualize>true</visualize>
          <update_rate>60.0</update_rate>
          <camera>
              <horizontal_fov>1.047198</horizontal_fov>
              <image>
                  <width>640</width>
                  <height>480</height>
                  <format>R8G8B8</format>
              </image>
              <clip>
                  <near>0.05</near>
                  <far>10</far>
              </clip>
          </camera>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
              <baseline>0.2</baseline>
              <alwaysOn>true</alwaysOn>
              <updateRate>0.0</updateRate>
              <frame_name>camera_frame</frame_name>
              <pointCloudCutoff>0.2</pointCloudCutoff>
              <pointCloudCutoffMax>10.0</pointCloudCutoffMax>
              <distortionK1>0</distortionK1>
              <distortionK2>0</distortionK2>
              <distortionK3>0</distortionK3>
              <distortionT1>0</distortionT1>
              <distortionT2>0</distortionT2>
              <CxPrime>0</CxPrime>
              <Cx>0</Cx>
              <Cy>0</Cy>
              <focalLength>0</focalLength>
              <hackBaseline>0</hackBaseline>
          </plugin> 
      </sensor> -->
  </gazebo>
  <gazebo reference="link1">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link1">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="link2">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link2">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="link3">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link3">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="link4">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link4">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="link5">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link5">
    <material>Gazebo/Gray</material>
  </gazebo>
  <gazebo reference="link6">
    <!-- <kp>1000000000000000000000000000.0</kp>
    <kd>1000000000000000000000000000.0</kd>
    <mu1>10.0</mu1>
    <mu2>10.0</mu2> -->
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link6">
    <material>Gazebo/Gray</material>
  </gazebo>
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>
  <material name="grey">
    <color rgba="0.5 0.5 0.5 1.0"/>
  </material>
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint4">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor4">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint5">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint6">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor6">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- World -->
  <link name="world"/>
  <!-- World fixed joint-->
  <joint name="world_fixed" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="world"/>
    <child link="link1"/>
  </joint>
  <!-- link 1 -->
  <link name="link1">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link1.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.00023 0.00000 0.06223"/>
      <mass value="0.85644"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00121064" ixy="0.0" ixz="0.00001100" iyy="0.00120886" iyz="0.0" izz="0.00038103" /> -->
    </inertial>
  </link>
  <!-- Joint 1 -->
  <joint name="joint1" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 0.126"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000" lower="-1.57079632679" upper="1.57079632679" velocity="6.5"/>
    <!-- <limit velocity="6.5" effort="1000" lower="-3.141592653589793" upper="3.141592653589793" /> -->
    <dynamics damping="0.7"/>
  </joint>
  <!-- link 2 -->
  <link name="link2">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link2.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.00000 0.00080 0.03183"/>
      <mass value="0.94658"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00152547" ixy="0.0" ixz="0.0" iyy="0.00045954" iyz="-0.00001086" izz="0.00151740" /> -->
    </inertial>
  </link>
  <!-- Joint 2 -->
  <joint name="joint2" type="revolute">
    <parent link="link2"/>
    <child link="link3"/>
    <origin rpy="0 0 0" xyz="0.0 0.069 0.033"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000" lower="-1.57079632679" upper="1.57079632679" velocity="6.5"/>
    <dynamics damping="0.7"/>
  </joint>
  <!-- link 3 -->
  <link name="link3">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link3.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.01768 -0.06881 0.20524"/>
      <mass value="1.30260"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.01111021" ixy="0.00001331" ixz="-0.00134208" iyy="0.01046642" iyz="0.00016235" izz="0.00202081" /> -->
    </inertial>
  </link>
  <!-- Joint 3 -->
  <joint name="joint3" type="revolute">
    <parent link="link3"/>
    <child link="link4"/>
    <origin rpy="0 -0.78 0" xyz="0.03000 -0.01150 0.26400"/>
    <axis xyz="0 1 0"/>
    <!-- <limit velocity="6.5" effort="1000" lower="-1.57079632679" upper="1.57079632679" /> -->
    <limit effort="1000" lower="-1.57079632679" upper="2.356194490192345" velocity="6.5"/>
    <dynamics damping="0.7"/>
  </joint>
  <!-- link 4 -->
  <link name="link4">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link4.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.12475 -0.05716 0.02979"/>
      <mass value="1.15977"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00096726" ixy="0.00004383" ixz="-0.00002771" iyy="0.00318801" iyz="0.00000627" izz="0.00341484" /> -->
    </inertial>
  </link>
  <!-- Joint 4 -->
  <joint name="joint4" type="revolute">
    <parent link="link4"/>
    <child link="link5"/>
    <origin rpy="-1.57 0 0" xyz="0.19500 -0.05750 0.03000"/>
    <axis xyz="1 0 0"/>
    <limit effort="1000" lower="-1.57079632679" upper="1.57079632679" velocity="6.5"/>
    <!-- <limit velocity="6.5" effort="1000" lower="-3.141592653589793" upper="3.141592653589793" /> -->
    <dynamics damping="0.7"/>
  </joint>
  <!-- link 5 -->
  <link name="link5">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link5.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.05642 -0.00217 0.00002"/>
      <mass value="0.44688"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00030285" ixy="0.00000761" ixz="0.00000031" iyy="0.00022339" iyz="-0.00000001" izz="0.00039766" /> -->
    </inertial>
  </link>
  <!-- Joint 5 -->
  <joint name="joint5" type="revolute">
    <parent link="link5"/>
    <child link="link6"/>
    <origin rpy="0 0 0" xyz="0.06300 0.04500 0.00000"/>
    <axis xyz="0 1 0"/>
    <limit effort="1000" lower="-1.57079632679" upper="1.57079632679" velocity="6.5"/>
    <dynamics damping="0.7"/>
  </joint>
  <!-- link 6 -->
  <link name="link6">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link6.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="link6.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.06831 -0.04457 0.00000"/>
      <mass value="0.43273"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00021512" ixy="0.00001009" ixz="0.00000000" iyy="0.00042827" iyz="0.00000000" izz="0.00051436" /> -->
    </inertial>
  </link>
  <!-- Joint 6 -->
  <joint name="joint6" type="revolute">
    <parent link="link6"/>
    <child link="end_link"/>
    <origin rpy="1.0472 0 0" xyz="0.12300 -0.04500 0.00000"/>
    <axis xyz="1 0 0"/>
    <limit effort="1000" lower="-1.57079632679" upper="1.57079632679" velocity="6.5"/>
    <!-- <limit velocity="6.5" effort="1000" lower="-3.141592653589793" upper="3.141592653589793" /> -->
    <dynamics damping="0.7"/>
  </joint>
  <!-- end link -->
  <link name="end_link">
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="end_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="end_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0.00337 0.00000 0.00000"/>
      <mass value="0.01919"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
      <!-- <inertia ixx="0.00000938" ixy="0.0" ixz="0.0" iyy="0.00000338" iyz="0.0" izz="0.00000635" /> -->
    </inertial>
  </link>
  <!-- CAMERA -->
  <joint name="camera_joint" type="fixed">
    <parent link="link6"/>
    <child link="camera_link"/>
    <origin xyz="0.09 -0.055 0.1" rpy="0 0 0"/>
  </joint>
  <link name="camera_link">
    <visual>
      <geometry>
        <box size="0.010 0.03 0.03"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="1e-5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.010 0.03 0.03"/>
      </geometry>
    </collision>
  </link>
  <joint name="camera_optical_joint" type="fixed">
    <parent link="camera_link"/>
    <child link="camera_link_optical"/>
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
  </joint>
  <link name="camera_link_optical"/>
  <!-- gripper main link -->
  <gazebo reference="gripper_main_link">
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <!-- <mu1>0.2</mu1>
    <mu2>0.2</mu2> -->
    <material>Gazebo/White</material>
  </gazebo>
  <!-- gripper link -->
  <gazebo reference="gripper_link">
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <!-- <mu1>0.2</mu1>
    <mu2>0.2</mu2> -->
    <material>Gazebo/White</material>
    <minDepth>0.003</minDepth>
    <maxVel>0</maxVel>
  </gazebo>
  <!-- gripper 1 link -->
  <gazebo reference="gripper_sub_link">
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
    <!-- <mu1>0.2</mu1>
    <mu2>0.2</mu2> -->
    <material>Gazebo/White</material>
    <minDepth>0.003</minDepth>
    <maxVel>0</maxVel>
  </gazebo>
  <!-- End Effector -->
  <!-- <gazebo reference="end_effector">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
    <material>Gazebo/Red</material>
  </gazebo> -->
  <joint name="base_link_gripper" type="fixed">
    <origin rpy="0 0 0" xyz="0.0 0 0"/>
    <parent link="end_link"/>
    <child link="gripper_main_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <!-- gripper main link -->
  <link name="gripper_main_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_main_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_main_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <mass value="0.509"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <!-- gripper joint -->
  <joint name="gripper" type="revolute">
    <axis xyz="0 0 -1"/>
    <origin rpy="0 0 0.7853981633974483" xyz="0.0552 -0.0460 0"/>
    <!-- <origin xyz="0.0552 -0.0370 0" rpy="0 0 0"/> -->
    <parent link="gripper_main_link"/>
    <child link="gripper_link"/>
    <limit effort="1000" lower="0.0" upper="1.1351457830353744" velocity="6.5"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <transmission name="tran_gripper">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gripper">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="motor_gripper">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface> -->
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- gripper link -->
  <link name="gripper_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.013"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <!-- gripper joint 1 -->
  <joint name="gripper_sub" type="fixed">
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 -0.7853981633974483" xyz="0.0552 0.0460 0"/>
    <!-- <origin xyz="0.0552 0.0370 0" rpy="0 0 0"/> -->
    <parent link="gripper_main_link"/>
    <child link="gripper_sub_link"/>
    <limit effort="1000" lower="0.0" upper="1.1351457830353744" velocity="6.5"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <transmission name="tran_gripper_sub">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gripper_sub">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="motor_gripper_sub">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface> -->
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- gripper 1 link -->
  <link name="gripper_sub_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_sub_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="gripper_sub_link.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.013"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <!-- gripper main fixed joint-->
  <!-- <joint name="gripper_main_fixed" type="fixed">
            <parent link="gripper_main_link"/>
            <child link="end_effector"/>
            <origin xyz="0.14 0.0 0.0" rpy="0 0 0" />
        </joint> -->
  <!-- End Effector -->
  <!-- <link name="end_effector">
            <visual>
            <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
            <geometry>      
                <box size = "0.01 0.01 0.01"/>
            </geometry>
            <material name="red"/>
            </visual>

            <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
            <mass value="0.001" /> 
            <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
            </inertial>
        </link> -->
  <ros2_control name="gripper_control" type="system">
    <hardware>
      <plugin>dynamixel_hardware/DynamixelHardware</plugin>
      <param name="usb_port">/dev/ttyUSB0</param>
      <param name="baud_rate">1000000</param>
    </hardware>
    <joint name="gripper">
      <param name="id">10</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0.0</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>
  <ros2_control name="openMANIPULATOR-P" type="system">
    <hardware>
      <plugin>dynamixel_hardware/DynamixelHardware</plugin>
      <param name="usb_port">/dev/ttyUSB0</param>
      <param name="baud_rate">1000000</param>
    </hardware>
    <joint name="joint1">
      <param name="id">1</param>
      <command_interface name="position">
        <param name="min">-0.2</param>
        <param name="max">0.2</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0.0</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint2">
      <param name="id">2</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0.0</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint3">
      <param name="id">3</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">0.74</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint4">
      <param name="id">4</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">1.57</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint5">
      <param name="id">5</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">1.21</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint6">
      <param name="id">6</param>
      <command_interface name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </command_interface>
      <command_interface name="velocity"/>
      <state_interface name="position">
        <param name="initial_value">-0.97</param>
      </state_interface>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
  </ros2_control>
</robot>
